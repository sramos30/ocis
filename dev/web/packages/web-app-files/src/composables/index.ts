export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './sort'

declare module 'vue/types/vue' {
  interface Vue {
    [key: string]: any
  }
}
