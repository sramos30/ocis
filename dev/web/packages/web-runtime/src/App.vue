<template>
  <div id="web">
    <oc-hidden-announcer :announcement="announcement" level="polite" />
    <skip-to target="web-content-main">
      <translate>Skip to main</translate>
    </skip-to>
    <component :is="layout"></component>
    <portal-target name="app.runtime.modal" multiple></portal-target>
    <oc-modal
      v-if="modal.displayed"
      :variation="modal.variation"
      :icon="modal.icon"
      :title="modal.title"
      :message="modal.message"
      :has-input="modal.hasInput"
      :input-label="modal.inputLabel"
      :input-disabled="modal.inputDisabled"
      :input-value="modal.inputValue"
      :input-description="modal.inputDescription"
      :input-error="modal.inputError"
      :button-cancel-text="modal.cancelText"
      :button-confirm-text="modal.confirmText"
      :button-confirm-disabled="modal.confirmDisabled || !!modal.inputError"
      @cancel="modal.onCancel"
      @confirm="modal.onConfirm"
      @input="modal.onInput"
      @mounted="focusModal"
      @beforeDestroy="focusModal"
    />
  </div>
</template>
<script lang="ts">
import { mapGetters, mapState, mapActions } from 'vuex'
import SkipTo from './components/SkipTo.vue'
import LayoutApplication from './layouts/Application.vue'
import LayoutLoading from './layouts/Loading.vue'
import LayoutPlain from './layouts/Plain.vue'
import { getBackendVersion, getWebVersion } from './container/versions'
import { defineComponent } from '@vue/composition-api'

export default defineComponent({
  components: {
    SkipTo
  },
  data() {
    return {
      $_notificationsInterval: null,
      windowWidth: 0,
      announcement: '',
      activeBlobStyle: {}
    }
  },
  computed: {
    ...mapState(['route', 'user', 'modal', 'sidebar']),
    ...mapGetters(['configuration', 'capabilities', 'getSettingsValue']),
    layout() {
      if (this.user.isAuthenticated && !this.user.userReady) {
        return LayoutLoading
      }

      if (
        !this.$route.name ||
        [
          'login',
          'oidcCallback',
          'oidcSilentRedirect',
          'privateLink',
          'publicLink',
          'accessDenied'
        ].includes(this.$route.name)
      ) {
        return LayoutPlain
      }

      return LayoutApplication
    },
    favicon() {
      return this.configuration.currentTheme.logo.favicon
    },

    selectedLanguage() {
      return this.getSettingsValue({
        extension: 'ocis-accounts',
        bundle: 'profile',
        setting: 'language'
      })
    }
  },
  watch: {
    $route: {
      immediate: true,
      handler: function (to) {
        this.announceRouteChange(to)
        document.title = this.extractPageTitleFromRoute(to)
      }
    },
    capabilities: {
      immediate: true,
      handler: function (caps) {
        if (!caps?.notifications) {
          return
        }

        // setup periodic loading of notifications if the server supports them
        this.$nextTick(() => {
          this.$_updateNotifications()
        })
        this.$_notificationsInterval = setInterval(() => {
          this.$_updateNotifications()
        }, 30000)
      }
    },
    selectedLanguage: {
      immediate: true,
      handler(language) {
        let languageCode = this.$language.defaultLanguage
        if (language !== null && language.listValue.values.length > 0) {
          languageCode = language.listValue.values[0].stringValue
        }
        if (languageCode) {
          this.$language.current = languageCode
          document.documentElement.lang = languageCode
        }
      }
    }
  },

  destroyed() {
    if (this.$_notificationsInterval) {
      clearInterval(this.$_notificationsInterval)
    }
  },

  metaInfo() {
    const metaInfo: any = {}
    if (this.favicon) {
      metaInfo.link = [{ rel: 'icon', href: this.favicon }]
    }
    const metaGenerator = {
      name: 'generator',
      content: [getWebVersion(), getBackendVersion({ store: this.$store })]
        .filter(Boolean)
        .join(', ')
    }
    metaInfo.meta = [metaGenerator]
    return metaInfo
  },

  methods: {
    ...mapActions(['fetchNotifications']),

    focusModal(component, event) {
      this.focus({
        revert: event === 'beforeDestroy'
      })
    },

    $_updateNotifications() {
      this.fetchNotifications(this.$client).catch((error) => {
        console.error('Error while loading notifications: ', error)
        clearInterval(this.$_notificationsInterval)
      })
    },

    announceRouteChange(route) {
      const pageTitle = this.extractPageTitleFromRoute(route, false)
      const translated = this.$gettext('Navigated to %{ pageTitle }')
      this.announcement = this.$gettextInterpolate(translated, { pageTitle })
    },

    extractPageTitleFromRoute(route, includeGeneralName = true) {
      const routeTitle = route.meta.title ? this.$gettext(route.meta.title) : route.name
      const titleSegments = [routeTitle]

      if (includeGeneralName) {
        titleSegments.push(this.configuration.currentTheme.general.name)
      }

      if (route.params.item) {
        if (route.name.startsWith('files-')) {
          const fileTree = route.params.item.split('/').filter((el) => el.length)

          if (fileTree.length) {
            titleSegments.unshift(fileTree.pop())
          }
        } else {
          titleSegments.unshift(route.params.item)
        }
      }

      return titleSegments.join(' - ')
    }
  }
})
</script>
<style lang="scss">
body {
  margin: 0;
}

#web {
  background-color: var(--oc-color-swatch-brand-default);
  height: 100vh;
  max-height: 100vh;
  overflow-y: hidden;
}

iframe {
  border: 0;
}
</style>
