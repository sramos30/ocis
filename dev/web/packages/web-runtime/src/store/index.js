import VuexPersistence from 'vuex-persist'

/* STORE MODULES
 */
import app from './app'
import apps from './apps'
import config from './config'
import user from './user'
import router from './router'
import settings from './settings'
import modal from './modal'
import navigation from './navigation'

const vuexPersistInSession = new VuexPersistence({
  key: 'webStateInSessionStorage',
  // Browser tab independent storage which gets deleted after the tab is closed
  storage: window.sessionStorage,
  reducer: (state) => {
    const { user, navigation, router } = state
    return {
      user,
      navigation,
      router
    }
  }
})

const strict = process.env.NODE_ENV === 'development'

export default {
  plugins: [vuexPersistInSession.plugin],
  modules: {
    app,
    apps,
    user,
    config,
    router,
    settings,
    modal,
    navigation
  },
  strict
}
