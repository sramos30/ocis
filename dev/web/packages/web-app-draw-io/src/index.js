import translations from '../l10n/translations'
import App from './App.vue'

const routes = [
  {
    name: 'draw-io',
    path: '/:filePath*',
    component: App,
    meta: {
      auth: false,
      patchCleanPath: true
    }
  }
]

const appInfo = {
  name: 'Draw.io',
  id: 'draw-io',
  icon: 'grid',
  extensions: [
    {
      extension: 'drawio',
      newTab: true,
      routeName: 'draw-io',
      newFileMenu: {
        menuTitle($gettext) {
          return $gettext('Draw.io document')
        }
      }
    },
    {
      extension: 'vsdx',
      newTab: true,
      routeName: 'draw-io'
    }
  ]
}

export default {
  appInfo,
  routes,
  translations
}
