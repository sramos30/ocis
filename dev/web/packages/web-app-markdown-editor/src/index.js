import App from './App.vue'
import translations from '../l10n/translations'

const routes = [
  {
    path: '/:filePath*',
    component: App,
    name: 'markdown-editor',
    meta: {
      patchCleanPath: true
    }
  }
]

const fileExtensionConfig = {
  canBeDefault: true
}

const appInfo = {
  name: 'MarkdownEditor',
  id: 'markdown-editor',
  icon: 'file-text',
  isFileEditor: true,
  extensions: [
    {
      extension: 'txt',
      newFileMenu: {
        menuTitle($gettext) {
          return $gettext('Plain text file')
        }
      },
      ...fileExtensionConfig
    },
    {
      extension: 'md',
      newFileMenu: {
        menuTitle($gettext) {
          return $gettext('Markdown file')
        }
      },
      ...fileExtensionConfig
    }
  ]
}

export default {
  appInfo,
  routes,
  translations
}
