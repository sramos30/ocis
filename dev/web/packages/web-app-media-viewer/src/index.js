import translations from '../l10n/translations'
import App from './App.vue'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const routes = [
  {
    path: '/:filePath*',
    component: App,
    name: 'media',
    meta: {
      auth: false,
      title: $gettext('Mediaviewer app'),
      patchCleanPath: true
    }
  }
]

const routeName = 'mediaviewer-media'

const fileExtensionConfig = {
  canBeDefault: true,
  routeName
}

const appInfo = {
  name: 'Mediaviewer',
  id: 'mediaviewer',
  icon: 'image',
  extensions: [
    {
      extension: 'png',
      ...fileExtensionConfig
    },
    {
      extension: 'jpg',
      ...fileExtensionConfig
    },
    {
      extension: 'jpeg',
      ...fileExtensionConfig
    },
    {
      extension: 'gif',
      ...fileExtensionConfig
    },
    {
      extension: 'mp4',
      ...fileExtensionConfig
    },
    {
      extension: 'webm',
      ...fileExtensionConfig
    },
    {
      extension: 'ogg',
      ...fileExtensionConfig
    }
  ]
}

export default {
  appInfo,
  routes,
  translations
}
