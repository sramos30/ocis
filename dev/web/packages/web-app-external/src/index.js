import translations from '../l10n/translations'
import App from './App.vue'
import store from './store'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: 'External',
  id: 'external'
}

const routes = [
  {
    name: 'apps',
    path: '/:fileId/:appName?',
    component: App,
    meta: {
      auth: false,
      title: $gettext('External app')
    }
  }
]

export default {
  appInfo,
  routes,
  store,
  translations,
  userReady({ store }) {
    store.dispatch('External/fetchMimeTypes')
  }
}
