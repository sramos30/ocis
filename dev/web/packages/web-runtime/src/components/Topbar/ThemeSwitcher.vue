<template>
  <oc-button
    v-oc-tooltip="buttonLabel"
    class="themeswitcher-btn"
    :aria-label="buttonLabel"
    appearance="raw"
    variation="inverse"
    @click="toggleTheme"
  >
    <span class="oc-visible@s" :aria-label="switchLabel" v-text="switchText" />
    <oc-icon :name="switchIcon" fill-type="line" />
  </oc-button>
</template>
<script>
import { computed, unref, watch } from '@vue/composition-api'
import { mapGetters } from 'vuex'
import { useStore, useLocalStorage } from 'web-pkg/src/composables'
import { themeNameDark, themeNameLight, useDefaultThemeName } from '../../composables'

export default {
  setup() {
    const store = useStore()
    const currentThemeName = useLocalStorage('oc_currentThemeName', useDefaultThemeName())
    const currentTheme = computed(() => store.getters.configuration.themes[unref(currentThemeName)])
    const applyTheme = (theme) => {
      for (const param in theme.designTokens.colorPalette) {
        document
          .querySelector(':root')
          .style.setProperty(`--oc-color-${param}`, theme.designTokens.colorPalette[param])
      }
    }

    watch(currentThemeName, async () => {
      await store.dispatch('loadTheme', { theme: unref(currentTheme) })
      applyTheme(unref(currentTheme))
    })

    return { currentThemeName, currentTheme }
  },
  computed: {
    ...mapGetters(['configuration']),
    isLightTheme() {
      return [null, themeNameLight].includes(this.currentThemeName)
    },
    buttonLabel() {
      return this.$gettext('Click to switch theme')
    },
    switchIcon() {
      return this.isLightTheme ? 'sun' : 'moon-clear'
    },
    switchLabel() {
      return this.$gettext('Currently used theme')
    },
    switchText() {
      return this.isLightTheme ? this.$gettext('Light mode') : this.$gettext('Dark mode')
    }
  },
  methods: {
    toggleTheme() {
      this.currentThemeName = this.isLightTheme ? themeNameDark : themeNameLight
    }
  }
}
</script>
