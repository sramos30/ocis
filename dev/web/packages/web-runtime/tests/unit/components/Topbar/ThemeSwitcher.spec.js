import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import GetTextPlugin from 'vue-gettext'
import DesignSystem from 'owncloud-design-system'
import CompositionApi from '@vue/composition-api'

import ThemeSwitcher from 'web-runtime/src/components/Topbar/ThemeSwitcher.vue'
import stubs from '../../../../../../tests/unit/stubs'
import { themeNameDark, themeNameLight } from '../../../../src/composables'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})
localVue.use(CompositionApi)

const lightTheme = {
  designTokens: {
    colorPalette: {
      'background-accentuate': 'lime'
    }
  }
}

const darkTheme = {
  designTokens: {
    colorPalette: {
      'background-accentuate': 'gold'
    }
  }
}

const spyToggleTheme = jest.spyOn(ThemeSwitcher.methods, 'toggleTheme')

describe('ThemeSwitcher component', () => {
  describe('visually', () => {
    beforeEach(() => {
      mockDarkModePreferred(false)
    })
    afterEach(() => {
      jest.clearAllMocks()
    })

    it('renders a button, initially in light mode per default', async () => {
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })

    it('renders a button, initially in dark mode if user prefers dark color scheme', async () => {
      mockDarkModePreferred(true)
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })

    it('toggles between themes upon click', async () => {
      const wrapper = getWrapper()
      expect(spyToggleTheme).toHaveBeenCalledTimes(0)

      await wrapper.find('.themeswitcher-btn').trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
      expect(spyToggleTheme).toHaveBeenCalledTimes(1)

      await wrapper.find('.themeswitcher-btn').trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
      expect(spyToggleTheme).toHaveBeenCalledTimes(2)
    })
  })

  describe('restores', () => {
    it('light theme if light theme is saved in localstorage', async () => {
      window.localStorage.setItem('oc_currentThemeName', themeNameLight)
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })
    it('dark theme if dark theme is saved in localstorage', async () => {
      window.localStorage.setItem('oc_currentThemeName', themeNameDark)
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function mockDarkModePreferred(enabled = false) {
  // matchMedia is not implemented in JSDOM, needs to be mocked
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => ({
      matches: enabled,
      media: "'(prefers-color-scheme: dark)'",
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn()
    }))
  })
}

function getWrapper() {
  return mount(ThemeSwitcher, {
    store: new Vuex.Store({
      getters: {
        configuration: () => ({
          themes: {
            default: lightTheme,
            'default-dark': darkTheme
          },
          currentTheme: lightTheme
        })
      },
      actions: {
        loadTheme: jest.fn()
      }
    }),
    localVue,
    stubs: { ...stubs, 'oc-button': false },
    directives: {
      'oc-tooltip': jest.fn()
    }
  })
}
