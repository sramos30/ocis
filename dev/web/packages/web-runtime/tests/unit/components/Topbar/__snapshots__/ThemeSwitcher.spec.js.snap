// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher component restores dark theme if dark theme is saved in localstorage 1`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component restores light theme if light theme is saved in localstorage 1`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="sun" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually renders a button, initially in dark mode if user prefers dark color scheme 1`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually renders a button, initially in light mode per default 1`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="sun" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually toggles between themes upon click 1`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually toggles between themes upon click 2`] = `
<button aria-label="Click to switch theme" type="button" class="themeswitcher-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="sun" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
</button>
`;
